### PR TITLE
make tests pass with enable_first_class_module() enabled.

### PR DESCRIPTION
**Pull request by zdevito**
Grabbag of fixes required to make it possible to pass the test suite using first-class module execution:

* Fixed a bug where Objects were not considered requires_grad when one of their members might be. The default requires_grad now just checks contained types for requires grad, which should be more robust.
* Edited some tests to be less fragile by removing checks for implementation details and putting some tests on more standard interfaces.
* Changed quantization so that it can find the parameters regardless of first-class mode or not.
* Provided `_lower_graph()` call that can be used by ONNX to still get the lowered exportable graph.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21565 [WIP] make tests pass with enable_first_class_module() enabled.**
* #21515 Add WeakIValue, use in tracer.

Differential Revision: [D15729503](https://our.internmc.facebook.com/intern/diff/D15729503)

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -10083,7 +10079,7 @@ def traced_fn(x):
         FileCheck().check("aten::neg").check("aten::add").run(str(traced_fn.graph))
 
     def test_call_script_mod_from_tracing_fn(self):
-        with self.disableEmitHook():
+        with self.assertRaisesRegex(RuntimeError, "must be registered as submodules"):
```
Comment (ZolotukhinM) [inline test/test_jit.py]: `FileCheck` call on line 10101 should probably be removed with this change.

File: torch/csrc/jit/passes/quantization.cpp [hunk 2]
```diff
@@ -48,59 +48,56 @@ Node* traverseToQuantNode(Node* dq) {
...
-  }
-  auto& opargs = optr->schema().arguments();
-  for (size_t idx = 0; idx < opargs.size(); idx++) {
-    if (opargs[idx].name() == param_name) {
-      return idx;
+struct ParamValue {
+  Value* definition;
+  IValue value;
+};
+static void gatherParamsFirstClass(script::Module& module, Value* module_value, std::vector<ParamValue>& params) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: Clang-format please 😛

File: torch/csrc/jit/python_tracer.cpp [hunk 3]
```diff
@@ -79,7 +79,7 @@ std::shared_ptr<torch::jit::Graph> createGraphByTracing(
           "captured in traces, so it would be a no-op.");
     }
     tracer::exit({toIValue(out)});
-    if (!script::getFirstClassMode()) {
+    if (true || !script::getFirstClassMode()) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/python_tracer.cpp]: Is this some leftover?

File: torch/csrc/jit/script/compiler.cpp [hunk 4]
```diff
@@ -3073,7 +3073,7 @@ void runCleanupPasses(std::shared_ptr<Graph>& to_clean, bool convert_ssa) {
   // be inlined into forked closures
   liftClosures(to_clean);
   inlineForkedClosures(to_clean);
-  if (!script::getFirstClassMode()) {
+  if (!script::getFirstClassMode() || true) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/script/compiler.cpp]: And this.

File: torch/csrc/jit/script/module.h [hunk 5]
```diff
@@ -99,6 +99,8 @@ struct TORCH_API Method {
     return *function_;
   }
 
+  std::pair<std::shared_ptr<Graph>, std::vector<at::Tensor>> _lowered_graph();
```
Comment (ZolotukhinM) [inline torch/csrc/jit/script/module.h]: This needs some documentation.

**Review by ZolotukhinM — APPROVED** [review]
I'd prefer this to be splitted into several PRs, since some of the changes are completely independent - e.g. quantization and `requires_grad` changes. And you already proposed a good separation in the description. Apart from that, the changes look good with just some minor comments (inline).

File: torch/csrc/jit/python_tracer.cpp (hunk 3 above)
Comment (zdevito) [reply torch/csrc/jit/python_tracer.cpp]: yeah, Ill remove it

**Pull request closed without merging**